rlmolecule/mcts/mcts.py

```diff
@@ -0,0 +1,173 @@
+import math
+import random
+from typing import List, Callable, Generator, Optional
+
+import numpy as np
+
+from rlmolecule.mcts.mcts_problem import MCTSProblem
+from rlmolecule.mcts.mcts_vertex import MCTSVertex
+from rlmolecule.tree_search.graph_search import GraphSearch
+from rlmolecule.tree_search.graph_search_state import GraphSearchState
+
+
+class MCTS(GraphSearch[MCTSVertex]):
+    def __init__(
+            self,
+            problem: MCTSProblem,
+            num_mcts_samples: int = 20,
+            ucb_constant: float = math.sqrt(2),
+    ) -> None:
+        super().__init__()
+        self._problem: MCTSProblem = problem
+        self._num_mcts_samples: int = num_mcts_samples
+        self.ucb_constant: float = ucb_constant
+
+    @property
+    def problem(self) -> MCTSProblem:
+        return self._problem
+
+    def run(
+            self,
+            state: Optional[GraphSearchState] = None,
+            explore: bool = True,
+            num_mcts_samples: Optional[int] = None,
+            mcts_selection_function: Optional[Callable[[MCTSVertex], MCTSVertex]] = None,
+    ) -> Generator[MCTSVertex]:
+        vertex = self._get_root() if state is None else self.get_vertex_for_state(state)
+        action_selection_function = self.softmax_selection if explore else self.visit_selection
+        num_mcts_samples = self._num_mcts_samples if num_mcts_samples is None else num_mcts_samples
+
+        def ucb_selection(parent):
+            return max(parent.children, key=lambda child: self._ucb_score(parent, child))
+
+        mcts_selection_function = ucb_selection if mcts_selection_function is None else mcts_selection_function
+
+        # noinspection PyTypeChecker
+        yield from self.run_from_vertex(vertex, action_selection_function, num_mcts_samples, mcts_selection_function)
```

Comment:
Is the functionality we get from `run` vs. `run_from_vertex` just the ability to add a different ucb score function? I'd probably default to just having the one option, I'm not sure that's something we expect to happen much -- especially if we're not using that functionality within the package, e.g. for alphazero.